File: onebusaway-android/src/main/java/org/onebusaway/android/speed/GammaSpeedEstimator.java
/*
 * Copyright (C) 2024 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.speed;

import org.onebusaway.android.io.elements.ObaTripSchedule;

import java.util.List;

/**
 * Speed estimator using the H12 gamma distribution model.
 * Combines schedule speed with the most recent AVL-derived speed to produce
 * a gamma distribution over vehicle speed.
 */
public class GammaSpeedEstimator implements SpeedEstimator {

    private final ScheduleSpeedEstimator scheduleEstimator = new ScheduleSpeedEstimator();
    private GammaSpeedModel.GammaParams mLastGammaParams;
    private double mLastScheduleSpeed;

    @Override
    public Double estimateSpeed(String vehicleId, VehicleState state,
                                TripDataManager dataManager) {
        mLastGammaParams = null;
        mLastScheduleSpeed = 0;

        Double scheduleSpeed = scheduleEstimator.estimateSpeed(vehicleId, state, dataManager);
        double vSched = scheduleSpeed != null ? scheduleSpeed : 0;
        mLastScheduleSpeed = vSched;

        String tripId = state.getActiveTripId();
        if (isTripNotYetStarted(tripId, vSched, dataManager)) {
            return null;
        }

        double vPrev = computePreviousAvlSpeed(tripId, dataManager);

        GammaSpeedModel.GammaParams params = GammaSpeedModel.fromSpeeds(vSched, vPrev);
        if (params != null) {
            mLastGammaParams = params;
            return GammaSpeedModel.medianSpeedMps(params);
        }

        // Fall back to schedule speed
        return vSched > 0 ? vSched : null;
    }

    private boolean isTripNotYetStarted(String tripId, double vSched,
                                         TripDataManager dataManager) {
        if (tripId == null || vSched <= 0) return false;
        Long serviceDate = dataManager.getServiceDate(tripId);
        ObaTripSchedule schedule = dataManager.getSchedule(tripId);
        ObaTripSchedule.StopTime[] stopTimes = schedule != null
                ? schedule.getStopTimes() : null;
        if (serviceDate != null && stopTimes != null && stopTimes.length > 0) {
            long tripStartMs = serviceDate + stopTimes[0].getDepartureTime() * 1000L;
            return System.currentTimeMillis() < tripStartMs;
        }
        return false;
    }

    private double computePreviousAvlSpeed(String tripId,
                                            TripDataManager dataManager) {
        if (tripId == null) return 0;
        List<VehicleHistoryEntry> history = dataManager.getHistoryReadOnly(tripId);
        if (history.size() < 2) return 0;

        VehicleHistoryEntry newer = null;
        VehicleHistoryEntry older = null;
        for (int i = history.size() - 1; i >= 0; i--) {
            VehicleHistoryEntry e = history.get(i);
            if (e.getBestDistanceAlongTrip() != null
                    && e.getLastLocationUpdateTime() > 0) {
                if (newer == null) {
                    newer = e;
                } else {
                    older = e;
                    break;
                }
            }
        }
        if (older == null || newer == null) return 0;

        long dtMs = newer.getLastLocationUpdateTime() - older.getLastLocationUpdateTime();
        if (dtMs <= 0) return 0;

        double dd = newer.getBestDistanceAlongTrip() - older.getBestDistanceAlongTrip();
        return Math.max(0, dd / (dtMs / 1000.0));
    }

    @Override
    public GammaSpeedModel.GammaParams getLastGammaParams() {
        return mLastGammaParams;
    }

    @Override
    public double getLastScheduleSpeed() {
        return mLastScheduleSpeed;
    }

    @Override
    public void clearState() {
        mLastGammaParams = null;
        mLastScheduleSpeed = 0;
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/extrapolation/math/speed/GammaSpeedEstimator.kt
/*
 * Copyright (C) 2024-2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.extrapolation.math.speed

import org.onebusaway.android.extrapolation.data.TripDataManager
import org.onebusaway.android.io.elements.ObaTripSchedule
import org.onebusaway.android.io.elements.ObaTripStatus
import org.onebusaway.android.io.elements.bestDistanceAlongTrip

/**
 * Speed estimator using a zero-inflated gamma distribution model. Combines schedule speed with
 * the most recent AVL-derived speed to produce a gamma distribution over vehicle speed. Falls
 * back to a point estimate from the schedule if the gamma model can't be computed.
 */
class GammaSpeedEstimator(private val dataManager: TripDataManager) : SpeedEstimator {

    companion object {
        /** Maximum time horizon (ms) for which gamma speed estimates are considered valid. */
        const val MAX_HORIZON_MS = 15 * 60 * 1000L // 15 minutes
    }

    private val scheduleEstimator = ScheduleSpeedEstimator(dataManager)

    /** Per-trip cached factory, keyed by tripId. Invalidated when lastFixTime changes. */
    private data class CachedFactory(val lastFixTime: Long, val factory: SpeedDistributionFactory)
    private val factoryCache = HashMap<String, CachedFactory>()

    override fun estimateSpeed(tripId: String, queryTime: Long): SpeedEstimateResult {
        val lastState = dataManager.getLastState(tripId)
        val schedule = dataManager.getSchedule(tripId)
        return estimateSpeedCore(tripId, queryTime, lastState, schedule)
    }

    override fun estimateSpeed(tripId: String, queryTime: Long,
                               snapshot: TripDataManager.TripSnapshot): SpeedEstimateResult =
            estimateSpeedCore(tripId, queryTime, snapshot.lastState, snapshot.schedule)

    private fun estimateSpeedCore(tripId: String, queryTime: Long,
                                   lastState: ObaTripStatus?,
                                   schedule: ObaTripSchedule?): SpeedEstimateResult {
        if (lastState == null) return SpeedEstimateResult.Failure(
                SpeedEstimateError.InsufficientData("No AVL fixes for trip"))

        val lastFixTime = lastState.lastLocationUpdateTime
        val dtSeconds = (queryTime - lastFixTime) / 1000.0
        if (dtSeconds < 0) return SpeedEstimateResult.Failure(
                SpeedEstimateError.TimestampOutOfBounds("Query time is before last AVL fix"))
        if (dtSeconds * 1000 > MAX_HORIZON_MS) return SpeedEstimateResult.Failure(
                SpeedEstimateError.TimestampOutOfBounds("Query time exceeds max horizon"))

        val cached = factoryCache[tripId]
        val factory = if (cached != null && cached.lastFixTime == lastFixTime) {
            cached.factory
        } else {
            val fixes = dataManager.mostRecentAvlFixes(tripId).take(2).toList()
            val prevSpeed = computeAvlSpeed(fixes)
            val scheduleSpeed =
                    when (val result = scheduleEstimator.estimateSpeedCore(lastState, schedule, queryTime)) {
                        is SpeedEstimateResult.Failure -> return result
                        is SpeedEstimateResult.Success -> result.distribution.mean
                    }
            makeGammaProbDistribution(scheduleSpeed, prevSpeed).also {
                factoryCache[tripId] = CachedFactory(lastFixTime, it)
            }
        }

        return SpeedEstimateResult.Success(factory.at(dtSeconds))
    }

    /** Clears the per-trip factory cache. */
    override fun clearCache() = factoryCache.clear()

    /** Computes speed from two AVL fixes (distance / time). Returns null if fewer than 2 fixes. */
    private fun computeAvlSpeed(fixes: List<ObaTripStatus>): Double? {
        if (fixes.size < 2) return null
        val (newer, older) = fixes
        val newerDist = newer.bestDistanceAlongTrip ?: return null
        val olderDist = older.bestDistanceAlongTrip ?: return null
        val dtMs = newer.lastLocationUpdateTime - older.lastLocationUpdateTime
        val dd = newerDist - olderDist
        return if (dtMs > 0) maxOf(0.0, dd / (dtMs / 1000.0)) else null
    }
}
